lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs

```diff
@@ -0,0 +1,301 @@
+    VectorStorageType,
+};
+use segment::vector_storage::query::context_query::ContextPair;
+use segment::vector_storage::query::discovery_query::DiscoveryQuery;
+use segment::vector_storage::query::reco_query::RecoQuery;
+use serde_json::json;
+use tempfile::Builder;
+
+use crate::utils::path;
+
+const MAX_EXAMPLE_PAIRS: usize = 4;
+
+enum QueryVariant {
+    Nearest,
+    RecommendBestScore,
+    Discovery,
+}
+
+fn random_multi_vec_discovery_query<R: Rng + ?Sized>(
+    rnd: &mut R,
+    dim: usize,
+    num_vector_per_points: usize,
+) -> QueryVector {
+    let num_pairs: usize = rnd.gen_range(1..MAX_EXAMPLE_PAIRS);
+
+    let target = random_multi_vector(rnd, dim, num_vector_per_points).into();
+
+    let pairs = (0..num_pairs)
+        .map(|_| {
+            let positive = random_multi_vector(rnd, dim, num_vector_per_points).into();
+            let negative = random_multi_vector(rnd, dim, num_vector_per_points).into();
+            ContextPair { positive, negative }
+        })
+        .collect_vec();
+
+    DiscoveryQuery::new(target, pairs).into()
+}
+
+fn random_multi_vec_reco_query<R: Rng + ?Sized>(
+    rnd: &mut R,
+    dim: usize,
+    num_vector_per_points: usize,
+) -> QueryVector {
+    let num_examples: usize = rnd.gen_range(1..MAX_EXAMPLE_PAIRS);
+
+    let positive = (0..num_examples)
+        .map(|_| random_multi_vector(rnd, dim, num_vector_per_points).into())
+        .collect_vec();
+    let negative = (0..num_examples)
+        .map(|_| random_multi_vector(rnd, dim, num_vector_per_points).into())
+        .collect_vec();
+
+    RecoQuery::new(positive, negative).into()
+}
+
+fn random_multi_vec_query<R: Rng + ?Sized>(
+    variant: &QueryVariant,
+    rnd: &mut R,
+    dim: usize,
+    num_vector_per_points: usize,
+) -> QueryVector {
+    match variant {
+        QueryVariant::Nearest => random_multi_vector(rnd, dim, num_vector_per_points).into(),
+        QueryVariant::Discovery => {
+            random_multi_vec_discovery_query(rnd, dim, num_vector_per_points)
+        }
+        QueryVariant::RecommendBestScore => {
+            random_multi_vec_reco_query(rnd, dim, num_vector_per_points)
+        }
+    }
+}
+
+/// Check all cases with single vector per multi and several vectors per multi
+#[rstest]
+#[case::nearest_eq(QueryVariant::Nearest, 1, 32, 5)]
+#[case::nearest_multi(QueryVariant::Nearest, 3, 32, 20)]
+#[case::discovery_eq(QueryVariant::Discovery, 1, 128, 5)]
+#[case::discovery_multi(QueryVariant::Discovery, 3, 128, 10)]
+#[case::recommend_eq(QueryVariant::RecommendBestScore, 1, 64, 5)]
+#[case::recommend_multi(QueryVariant::RecommendBestScore, 2, 64, 10)]
+fn test_multi_filterable_hnsw(
+    #[case] query_variant: QueryVariant,
+    #[case] num_vector_per_points: usize,
+    #[case] ef: usize,
+    #[case] max_failures: usize, // out of 100
+) {
+    _test_multi_filterable_hnsw(query_variant, num_vector_per_points, ef, max_failures);
```

Comment:
Why do you need `_test_multi_filterable_hnsw` as separate function here? I propose to inline `_test_multi_filterable_hnsw`